poc_iot_verifier/src/rewarder.rs

```diff
@@ -0,0 +1,121 @@
+use crate::{reward_share::GatewayShares, scheduler::Scheduler, Result};
+use chrono::{DateTime, Duration, TimeZone, Utc};
+use db_store::meta;
+use file_store::{file_sink, file_sink_write, traits::TimestampEncode};
+use helium_proto::RewardManifest;
+use sqlx::{PgExecutor, Pool, Postgres};
+use tokio::time::sleep;
+
+pub struct Rewarder {
+    pub pool: Pool<Postgres>,
+    pub gateway_rewards_tx: file_sink::MessageSender,
+    pub reward_manifest_tx: file_sink::MessageSender,
+    pub reward_period_hours: i64,
+    pub reward_offset: Duration,
+}
+
+impl Rewarder {
+    pub async fn run(mut self, shutdown: &triggered::Listener) -> Result {
+        tracing::info!("Starting iot verifier rewarder");
+
+        let reward_period_length = Duration::hours(self.reward_period_hours);
+
+        loop {
+            let now = Utc::now();
+
+            let scheduler = Scheduler::new(
+                reward_period_length,
+                fetch_rewarded_timestamp("last_rewarded_end_time", &self.pool).await?,
+                fetch_rewarded_timestamp("next_rewarded_end_time", &self.pool).await?,
+                self.reward_offset,
+            );
+
+            if scheduler.should_reward(now) {
+                tracing::info!("Rewarding for period: {:?}", scheduler.reward_period);
+                self.reward(&scheduler).await?
+            }
+
+            let sleep_duration = scheduler.sleep_duration(Utc::now())?;
+
+            tracing::info!(
+                "Sleeping for {}",
+                humantime::format_duration(sleep_duration)
+            );
+
+            let shutdown = shutdown.clone();
+            tokio::select! {
+                _ = shutdown => return Ok(()),
+                _ = sleep(sleep_duration) => (),
+            }
+        }
+    }
+
+    pub async fn reward(&mut self, scheduler: &Scheduler) -> Result {
+        let reward_shares =
+            GatewayShares::aggregate(&self.pool, scheduler.reward_period.end).await?;
+        for reward_share in reward_shares.into_gateway_reward_shares(&scheduler.reward_period) {
+            file_sink_write!(
+                "gateway_reward_shares",
+                &self.gateway_rewards_tx,
+                reward_share
+            )
+            .await?
+            // Await the returned oneshot to ensure we wrote the file
+            .await??;
+        }
+
+        let written_files = file_sink::fetch_manifest(&self.gateway_rewards_tx)
+            .await?
+            .await??;
```

Comment:
what be this double wait malarkey  ?

Reply:
i commented the double-wait above (when file sinking the reward shares files) but i can comment this one too for clarity; the double-await is 1) await and "unwrap" the result of the file write operation which yields an oneshot receiver channel which 2) we await and "unwrap" the response from (the oneshot receiver response being a result which is itself wraps a result, hence the double-??)
basically we want to await the channel response to ensure that the file write/sink completed successfully